compiler/rustc_const_eval/src/interpret/memory.rs

```diff
@@ -1222,6 +1222,59 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
 
         Ok(())
     }
+
+    pub fn mem_swap_nonoverlapping(
+        &mut self,
+        x_ptr: Pointer<Option<M::Provenance>>,
+        y_ptr: Pointer<Option<M::Provenance>>,
+        count: u64,
+        layout: ty::layout::TyAndLayout<'tcx>,
+    ) -> InterpResult<'tcx> {
+        let elem_size = layout.size;
+        let align = layout.align.abi;
+
+        if count > i64::MAX as u64 {
+            throw_ub_format!("`count` argument to `swap_nonoverlapping_many` is too large.");
+        }
+
+        let first_ptr_acc = self.get_ptr_access(x_ptr, elem_size * count, align)?;
+        let second_ptr_acc = self.get_ptr_access(y_ptr, elem_size * count, align)?;
+
+        let Some((x_alloc_id, x_offset, _)) = first_ptr_acc else {
+            assert_eq!(elem_size, Size::ZERO);
+            // Called on ZST so it is noop.
+            return Ok(())
+        };
+        let Some((y_alloc_id, y_offset, _)) = second_ptr_acc else {
+            unreachable!("If right param is ZST, left must be too")
+        };
+
+        if x_alloc_id == y_alloc_id {
+            if (x_offset..x_offset + elem_size * count).contains(&y_offset)
+                || (y_offset..y_offset + elem_size * count).contains(&x_offset)
+            {
+                throw_ub_format!("swap was called on overlapping memory.");
+            }
+        }
+
+        if count == 0 {
+            return Ok(());
+        }
+
+        let tmp_stack_alloc = self.allocate(layout, MemoryKind::Stack)?;
```

Comment:
Making this 'stack' memory is not right. This special magic memory.^^ We'll probably need a new MemoryKind, `Scratch` or something like that.

Reply:
Are you suggesting to add this new MemoryKind?

Reply:
> Are you suggesting to add this new MemoryKind?

Yes.

> P.S. What is so special about MemoryKind::Stack?

It's not special, but it's for stack memory. This here is not stack memory, so we shouldn't give a wrong MemoryKind.